src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/oidc/OpenIDAuthentication.java

```diff
@@ -0,0 +1,46 @@
+package edu.harvard.iq.dataverse.authorization.providers.oauth2.oidc;
+
+import java.io.IOException;
+
+import edu.harvard.iq.dataverse.util.SystemConfig;
+import fish.payara.security.annotations.LogoutDefinition;
+import fish.payara.security.annotations.OpenIdAuthenticationDefinition;
+import fish.payara.security.openid.api.OpenIdConstant;
+import jakarta.annotation.security.DeclareRoles;
+import jakarta.ejb.EJB;
+import jakarta.servlet.ServletException;
+import jakarta.servlet.annotation.HttpConstraint;
+import jakarta.servlet.annotation.ServletSecurity;
+import jakarta.servlet.annotation.WebServlet;
+import jakarta.servlet.http.HttpServlet;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+
+/**
+ * OIDC login implementation
+ */
+@WebServlet("/oidc/login")
+@OpenIdAuthenticationDefinition(
+    providerURI = "#{openIdConfigBean.providerURI}",
+    clientId = "#{openIdConfigBean.clientId}",
+    clientSecret = "#{openIdConfigBean.clientSecret}",
+    redirectURI = "#{openIdConfigBean.redirectURI}",
+    logout = @LogoutDefinition(redirectURI = "#{openIdConfigBean.logoutURI}"),
+    scope = {OpenIdConstant.OPENID_SCOPE, OpenIdConstant.EMAIL_SCOPE, OpenIdConstant.PROFILE_SCOPE},
+    tokenAutoRefresh = true,
+    useSession = true // If enabled state & nonce value stored in session otherwise in cookies.
+)
+@DeclareRoles("all")
+@ServletSecurity(@HttpConstraint(rolesAllowed = "all"))
+public class OpenIDAuthentication extends HttpServlet {
+    @EJB
+    OpenIDConfigBean openIdConfigBean;
+
+    @Override
+    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
```

Comment:
I'm not sure why this works and how the pieces fit together. In `OidcLoginBackingBean.loginLink()` you generate a link to enable logging in to the provider as `/oidc/login/target=JSF&oidcp=...`. People clicking on this link usually should be visiting the IdP login page, not this endpoint. What is going on here? Could you explain a litte more please? Thx!

Reply:
For some reason we do not use security annotations at all. If we want to use the OIDC annotation, it assumes we also use the security annotations, http authentication mechanism and the identity store (which we do not use...). If I do not add security annotation to the authentication servlet, it will not be secured and will not force you to log in, you will just see its content, OIDC annotation has no effect. This is why you noticed the identity store annotation "hack". This is one of the "hacks" that makes this implementation work -> authentication servlet is the only location that is really secured with OIDC and security annotations, security context, authentication mechanism handler, http authentication mechanism and the bearer token mechanism. But we do not use it anywhere else (and I assume we do not want to), so this servlet has content that is unreachable (because we do not use security annotations anywhere else in authentication mechanisms we implemented, so you cannot gain access to it by any means). I renamed the required role to `nobodyHasAccess` in the hope that it would make things more clear:
```
/**
 * OIDC login implementation
 */
@WebServlet("/oidc/login")
@OpenIdAuthenticationDefinition(
    providerURI = "#{openIdConfigBean.providerURI}",
    clientId = "#{openIdConfigBean.clientId}",
    clientSecret = "#{openIdConfigBean.clientSecret}",
    redirectURI = "#{openIdConfigBean.redirectURI}",
    logout = @LogoutDefinition(redirectURI = "#{openIdConfigBean.logoutURI}"),
    scope = {OpenIdConstant.OPENID_SCOPE, OpenIdConstant.EMAIL_SCOPE, OpenIdConstant.PROFILE_SCOPE},
    useSession = true // If enabled state & nonce value stored in session otherwise in cookies.
)
@DeclareRoles("nobodyHasAccess")
@ServletSecurity(@HttpConstraint(rolesAllowed = "nobodyHasAccess"))
public class OpenIDAuthentication extends HttpServlet {
```
You can only make calls to the API with the bearer tokens in this implementation because the bearer token identity store passes the user to the AbstractApiBean, so it can just rely on the "standard" Dataverse implementation, without any security annotations. The authentication servlet is there only to force the use of the OIDC annotation to do its work, so it is not important that it is not reachable. After log in you are redirected somewhere else anyway.